### PR TITLE
Fix alibi

### PR DESCRIPTION
**Pull request by thomasw21**

### CODE REVIEW

File: megatron/model/transformer.py [hunk 1]
```diff
@@ -303,11 +303,27 @@ def forward(self, hidden_states, attention_mask, layer_past=None,
...
+                beta=0.0, alpha=(1.0/self.norm_factor))
+        else:
+            if not hasattr(self, "logged_alibi"):
+                logger.debug("Using Alibi.")
+                self.logged_alibi = True
+
+            if self.apply_query_key_layer_scaling:
+                beta = 1.0 / self.layer_number
+            else:
+                beta = 1.0
```
Comment (thomasw21) [inline megatron/model/transformer.py:320]: That's the side fix about scaling. The issue is the models apply scaling by a factor `self.layer_number` later on. In order to neutralise that scaling we change beta value.

File: megatron/model/transformer.py [hunk 2]
```diff
@@ -303,11 +303,27 @@ def forward(self, hidden_states, attention_mask, layer_past=None,
...
+        if alibi is None:
+            matmul_result = torch.baddbmm(
+                matmul_result,
+                query_layer.transpose(0, 1),   # [b * np, sq, hn]
+                key_layer.transpose(0, 1).transpose(1, 2),  # [b * np, hn, sk]
+                beta=0.0, alpha=(1.0/self.norm_factor))
+        else:
+            if not hasattr(self, "logged_alibi"):
+                logger.debug("Using Alibi.")
+                self.logged_alibi = True
```
Comment (thomasw21) [inline megatron/model/transformer.py:315]: Testing purposes.

Comment (SaulLu) [reply megatron/model/transformer.py:315]: Just for information, on my side I was [going to propose to isolate](https://github.com/bigscience-workshop/Megatron-DeepSpeed/commit/8cb55347a41b3063d69cf37986fd2cc630598258) the 3 ways to calculate the score according to the positional embedding by creating a method by way of calculating the score and then wrapping these methods in FF to have a log (as for the activation functions) to detect in the tests.

The micro advantage is that it also allows to test rotary and absolute (in all cases) but I don"t mind if you think it's easier to keep it like you did.

File: megatron/model/transformer.py (hunk 1 above)
Comment (SaulLu) [reply megatron/model/transformer.py:320]: Nice!

**Review by SaulLu — COMMENTED** [review]
Thanks a lot for the fix! :confetti_ball: 

Can we also add a test in `test_model.py` in order to test alibi when we use it without megatron?

**Comment by thomasw21** [comment]
Not super clear on what you mean by `without megatron`. The test in `test_model` tests out invariants that we can have using prefix lm or gpt. I don't see any invariants for alibi version?

File: megatron/model/transformer.py (hunk 2 above)
Comment (thomasw21) [reply megatron/model/transformer.py:315]: Yeah that's true that your version looks nice. I just think it's an incorrect abstraction. There's no reason to group all the positional embeddings together (they get applied in different places, they do different things, they have different constraints despite having the common purpose on given sequential information). One could argue that using a pure causal mask is a position embedding mechanism.

What I was thinking of is abstracting only the alibi function in a seperate function to use the pretty decorator, but I was lazy ^^' `@log_debug_usage(logger, msg)`

**Comment by SaulLu** [comment]
Oh sorry, I was meaning `without deepspeed` !

**Comment by thomasw21** [comment]
Ah makes sense ... I'll try something out

**Comment by thomasw21** [comment]
Okay after thinking about it. I think we need to create a test so that all the models we have match their deepspeed version. I don't think we should implement that test in this PR. I'll open an issue.

**Review by SaulLu — APPROVED** [review]
Following the last offline discussions, if ever we have to run the evaluation on the Megatron version without deepspeed we will add a test to verify that the architectures are identical.

In the meantime, I have full confidence in this fix and I think we can move forward and re-launch a training.

Thanks again Thomas! :rocket:

File: megatron/model/transformer.py (hunk 2 above)
Comment (ofirpress) [reply megatron/model/transformer.py:315]: > using a pure causal mask is a position embedding mechanism
I agree with this statement :) Models that don't have any position embeddings (like sinusoidal or learned or alibi) are actually able to achieve good (but not great) PPL because the causal mask encodes some kind of order.

File: megatron/model/transformer.py [hunk 3]
```diff
@@ -470,9 +486,19 @@ def __init__(self, init_method, output_layer_init_method,
         self.mlp = ParallelMLP(init_method,
                                output_layer_init_method)
 
+        # Alibi
+        if args.position_embedding_type == PositionEmbeddingType.alibi:
+            self.alibi = self._build_alibi_tensor(args.seq_length, args.num_attention_heads, args.micro_batch_size).to(torch.cuda.current_device())
```
Comment (DanielHesslow) [inline megatron/model/transformer.py:491]: Does `args.micro_batch_size` work with batch size rampup? And if not, do we care?

Comment (thomasw21) [reply megatron/model/transformer.py:491]: micro batch size doesn't increase during batch size rampup. It's constant.

We should care since all our experiment run with batch size rampup, but I would expect it to crash badly if it doesn't match.

**Comment by DanielHesslow** [comment]
Have we tested that we get the same results with/without tensor parallelism? Correct me if I'm wrong but I think that the attention heads are distributed across workers since self.query_key_value is ColumnParallel. We might need to extract the correct part of the alibi tensor depending on the tensor parallel index since different attention heads should have different alibi slopes?

ie. https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/33f43ee51e7ae2a333c05d4fb4c3d5bf13ec4ea2/megatron/model/transformer.py#L291
might be wrong?

**Comment by thomasw21** [comment]
That's actually a great point. Looking at things in diagonal, I feel you're completely right! Though we are lucky as we're using TP=1 in our experiment. If I understand correctly you believe `:output_size[0]*output_size[1]` is wrong right? because it doesn't take in account the ordering of the heads, ie there should be an offset linked to the `tp_rank` right?

If so I would suggest you open an issue/PR to introduce such test, and if it fails to fix it?

**Comment by DanielHesslow** [comment]
Yup exactly there should be some offset depending on the tp_rank, and possibly also a stride since the batch_size is baked in the first dimension as well. 

I've opened an issue for it here: #227
